rasa/core/policies/ensemble.py

```diff
@@ -91,590 +218,315 @@ def _check_for_important_policies(self) -> None:
+        final.optional_events = []
+        final.action_metadata = None
 
-            # End-to-end predictions overrule all other predictions based on user input.
-            if (
-                not is_no_user_prediction
-                and prediction.is_end_to_end_prediction != is_end_to_end_prediction
-            ):
-                continue
+        return final
 
-            confidence = (prediction.max_confidence, prediction.policy_priority)
-            if confidence > best_confidence:
-                # pick the best policy
-                best_confidence = confidence
-                best_policy_name = policy_name
+    def _choose_best_prediction(self, predictions: List[PolicyPrediction2]) -> int:
+        """Chooses the "best" prediction out of the given predictions.
 
-        best_prediction = predictions.get(best_policy_name)
-
-        if not best_prediction:
-            raise InvalidConfigException(
-                f"No prediction for policy '{best_policy_name}' found. Please check "
-                f"your model configuration."
-            )
-
-        policy_events += best_prediction.optional_events
-
-        return PolicyPrediction(
-            best_prediction.probabilities,
-            best_policy_name,
-            best_prediction.policy_priority,
-            policy_events,
-            is_end_to_end_prediction=best_prediction.is_end_to_end_prediction,
-            is_no_user_prediction=best_prediction.is_no_user_prediction,
-            diagnostic_data=best_prediction.diagnostic_data,
-            hide_rule_turn=best_prediction.hide_rule_turn,
-        )
-
-    def _best_policy_prediction(
-        self,
-        tracker: DialogueStateTracker,
-        domain: Domain,
-        interpreter: NaturalLanguageInterpreter,
-    ) -> PolicyPrediction:
-        """Finds the best policy prediction.
+        Note that this comparison is *not* symmetric if the priorities are allowed to
+        coincide (i.e. if we cannot distinguish two predictions using 1.-4., then
+        the first prediction is considered to be "better").
 
         Args:
-            tracker: the :class:`rasa.core.trackers.DialogueStateTracker`
-            domain: the :class:`rasa.shared.core.domain.Domain`
-            interpreter: Interpreter which may be used by the policies to create
-                additional features.
-
+          predictions: all predictions made by an ensemble of policies
         Returns:
-            The winning policy prediction.
+          index of the best prediction
         """
-        # find rejected action before running the policies
-        # because some of them might add events
-        rejected_action_name = None
-        last_action_event = next(
-            (
-                event
-                for event in reversed(tracker.events)
-                if isinstance(event, (ActionExecutionRejected, ActionExecuted))
-            ),
-            None,
+        does_not_contain_no_user_predictions = all(
```

Comment:
can we please remove this double negation? Makes the code in my opinion hard to understand

Reply:
maybe `contains_no_form_predictions`?

Reply:
agree - I just didn't properly understand how to rephrase "no user prediction"